velox/exec/tests/SparkShuffleTest.cpp

```diff
@@ -0,0 +1,611 @@
+                .build()};
+  }
+};
+
+class TestShuffle : public exec::Shuffle {
+ public:
+  TestShuffle(
+      memory::MemoryPool* pool,
+      uint32_t numPartitions,
+      uint32_t maxBytesPerPartition)
+      : pool_{pool},
+        numPartitions_{numPartitions},
+        maxBytesPerPartition_{maxBytesPerPartition},
+        inProgressSizes_(numPartitions, 0) {
+    inProgressPartitions_.resize(numPartitions_);
+    readyPartitions_.resize(numPartitions_);
+  }
+
+  void collect(int32_t partition, std::string_view data) {
+    auto& buffer = inProgressPartitions_[partition];
+
+    // Check if there is enough space in the buffer.
+    if (buffer &&
+        inProgressSizes_[partition] + data.size() + sizeof(size_t) >=
+            maxBytesPerPartition_) {
+      buffer->setSize(inProgressSizes_[partition]);
+      readyPartitions_[partition].emplace_back(std::move(buffer));
+      inProgressPartitions_[partition].reset();
+    }
+
+    // Allocate buffer if needed.
+    if (!buffer) {
+      buffer = AlignedBuffer::allocate<char>(maxBytesPerPartition_, pool_);
+      inProgressSizes_[partition] = 0;
+    }
+
+    // Copy data.
+    auto rawBuffer = buffer->asMutable<char>();
+    auto offset = inProgressSizes_[partition];
+
+    *(size_t*)(rawBuffer + offset) = data.size();
+
+    offset += sizeof(size_t);
+    memcpy(rawBuffer + offset, data.data(), data.size());
+
+    inProgressSizes_[partition] += sizeof(size_t) + data.size();
+  }
+
+  void noMoreData() {
+    for (auto i = 0; i < numPartitions_; ++i) {
+      if (inProgressSizes_[i] > 0) {
+        auto& buffer = inProgressPartitions_[i];
+        buffer->setSize(inProgressSizes_[i]);
+        readyPartitions_[i].emplace_back(std::move(buffer));
+        inProgressPartitions_[i].reset();
+      }
+    }
+  }
+
+  bool hasNext(int32_t partition) const {
+    return !readyPartitions_[partition].empty();
+  }
+
+  BufferPtr next(int32_t partition) {
+    VELOX_CHECK(!readyPartitions_[partition].empty());
+
+    auto buffer = readyPartitions_[partition].back();
+    readyPartitions_[partition].pop_back();
+    return buffer;
+  }
+
+ private:
+  memory::MemoryPool* pool_;
+  const uint32_t numPartitions_;
+  const uint32_t maxBytesPerPartition_;
+  std::vector<BufferPtr> inProgressPartitions_;
+  std::vector<size_t> inProgressSizes_;
+  std::vector<std::vector<BufferPtr>> readyPartitions_;
+};
+
+class UnsafeRowExchangeSource : public ExchangeSource {
+ public:
+  UnsafeRowExchangeSource(
+      const std::string& taskId,
+      int destination,
+      std::shared_ptr<exec::ExchangeQueue> queue,
+      Shuffle* shuffle,
+      memory::MemoryPool* pool)
+      : ExchangeSource(taskId, destination, queue, pool), shuffle_(shuffle) {}
+
+  bool shouldRequestLocked() override {
+    return !atEnd_;
+  }
+
+  void request() override {
+    std::lock_guard<std::mutex> l(queue_->mutex());
+
+    if (!shuffle_->hasNext(destination_)) {
+      atEnd_ = true;
+      queue_->enqueue(nullptr);
```

Comment:
What does inserting nullptr in the queue will lead to?

Reply:
This is a signal that there will be no more data coming.